osutil/vfs/propagation.go

```diff
@@ -0,0 +1,274 @@
+				if m.hasAncestor(ancestorID) {
+					fn(m)
+				}
+			}
+		}
+
+		return nil
+	})
+}
+
+// atMountPoint calls [fn] if [mountPoint] is an existing mount point.
+//
+// If [mountPoint] is not a mount point then a [fs.PathError] is returned,
+// encapsulating the given path [mountPoint], and operation [op].
+func (v *VFS) atMountPoint(mountPoint, op string, fn func(pd pathDominator) error) error {
+	// Find the mount dominating the mount point.
+	pd := v.pathDominator(mountPoint)
+
+	// If the VFS suffix is not empty then the given path is _not_ a mount point.
+	if pd.suffix != "" && pd.suffix != "." {
+		return &fs.PathError{Op: op, Path: mountPoint, Err: errNotMounted}
+	}
+
+	return fn(pd)
+}
+
+// sliceContains returns true if the given slice [sl] contains element [elem].
+func sliceContains[T comparable](sl []T, elem T) bool {
+	for i := range sl {
+		if sl[i] == elem {
+			return true
+		}
+	}
+
+	return false
+}
+
+// unlockedSharedGroupCount countns the members of the shared peer group [id].
+func (v *VFS) unlockedSharedGroupCount(id GroupID) (count int) {
+	for _, m := range v.mounts {
+		if m.shared == id {
+			count++
+		}
+	}
+
+	return count
+}
+
+// maybePropagateMount propagates the mount m among the members of the parent's
+// peer group and their slaves. The function does nothing if the parent mount
+// is not shared.
+func (v *VFS) propagateMount(m *mount, skipFn func(m *mount) bool) {
+	if m.parent == nil {
+		panic("m.parent == nil")
+	}
+
+	if m.parent.shared == 0 {
+		panic("m.parent.shared == 0")
+	}
+
+	// TODO: use range syntax once we upgrade to go 1.23
+	v.peersAndSlavesOf(m.parent.shared)(func(peer *mount) bool {
+		if skipFn(peer) {
+			return true
+		}
+
+		// Skip this peer if it doesn't see the part of the file system that
+		// the event originated from. Here m.attachedAt s the place where the
+		// event originated within the parent mount, and peer.rootDir is the
+		// sub-tree of the parent mount that is actually visible in the peer.
+		if !strings.HasPrefix(m.attachedAt, peer.rootDir) {
+			return true
+		}
+
+		// Re-locate the attachment point based on the sub-tree that the peer
+		// is using. Note that we just checked the presence of the prefix in
+		// the test above.
+		attachedAt := strings.TrimPrefix(m.attachedAt, peer.rootDir)
+
+		// Compute how the new propagated mount is shared based on the type of
+		// peer (peer or slave peer) is the target of the propagation.
+		var master, shared GroupID
+		if peer.shared == m.parent.shared {
+			// Propagating into a peer. The shared group is retained.
+			shared = m.shared
+		} else {
+			// Propagating into a slave. The shared group becomes slave.
+			master = m.shared
+			if shared == 0 && peer.shared != 0 {
+				// The slave is also shared (independently of the peer group
+				// that undergoes propagation) so the propagated mount joins a
+				// new peer group.
+				//
+				// FIXME(propagation): Should this propagate again? After all,
```

Comment:
This needs careful analysis of kernel behavior.

Reply:
This is slightly better and worse at the same time. We need to iterate over a larger set of nodes and the iterator is somewhat complex: https://elixir.bootlin.com/linux/v6.16/source/fs/pnode.c#L183 -- see https://elixir.bootlin.com/linux/v6.16/source/fs/pnode.c#L291 for usage.